### DISCUSSION TITLE
New markup format

### DISCUSSION THREAD

**Discussion by gyscos** [gyscos/cursive#797] [Ideas]
I'm thinking of adding support to a new minimal markup format meant to more easily generate `StyledString`.

I'd like it to be simpler than html `<red>This text<red> is written in red`.

I like the latex style of `\foo{bar}{baz}`. Unfortunately, the way to use colors in latex is actually a bit verbose: `\textcolor{red}{This text} is written in red` or `{\textcolor{red}This text} is written in red`. In addition, using backslashes would be annoying without raw strings, which puts limitations on inserting unicode characters by code. An alternative I've been toying with is `/red{This text} is written in red`. There could be a few variations to bikeshed that:

* `/red{This text} is written in red`
* `{This text}{red} is written in red`

In addition, to apply multiple styles, we could combine them like `/red+bold{warning}`. Maybe also allow setting the background color with `back=` or `back.`: 
* `/red+back=blue{foobar}`
* `/red+back.blue{foobar}`

The current `main` branch has a basic implementation for `/red+bold{foobar}` (no background color yet though). It's also currently not dealing well with nested styles like `/red{Read /bold{that}}` - at the moment it should be flattened to `/red{Read }/red+bold{that}`. That's just a limitation of the basic regex I'm using - I'll replace the implementation to be more robust (and maybe not need regex at all, so it would not be behind a feature flag).

**Comment by correabuscar** [comment]
For what is worth,
`cargo test --all-features -- utils::markup::cursup`
fails at this:
https://github.com/gyscos/cursive/commit/042a2f8c98c8ea22fbfde2129e7ef8fdc9a30324#diff-1015a5e1569edf0d190fd82601da788ca734e54d75d4ae363f5816e641d2681dR6-R7

like:
```rust
running 1 test
test cursive-core/src/utils/markup/cursup.rs - utils::markup::cursup (line 5) ... FAILED

failures:

---- cursive-core/src/utils/markup/cursup.rs - utils::markup::cursup (line 5) stdout ----
error: expected expression, found `/`
 --> cursive-core/src/utils/markup/cursup.rs:6:1
  |
3 | /red{This} /green{text} /blue{is} /bold{very} /underline{styled}!
  | ^ expected expression

error: aborting due to 1 previous error

Couldn't compile the test.

failures:
    cursive-core/src/utils/markup/cursup.rs - utils::markup::cursup (line 5)

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 142 filtered out; finished in 0.05s

error: doctest failed, to rerun pass `-p cursive_core --doc`
```


What's curious is, that none of these catch it:
`cargo test`
`cargo test --all-targets`
`cargo test --all-targets --all-features`
but this does:
`cargo test --all-features`

(which is why I've been using all 4 when checking for breakage, manually, locally)

I should probably mention that for `--all-features` to compile(for me?) I had to do a lil diff:
```diff
-[dependencies.bear-lib-terminal]
-optional = true
-version = "2"
+#[dependencies.bear-lib-terminal]
+#optional = true
+#version = "2"

-blt-backend = ["bear-lib-terminal"]
+#blt-backend = ["bear-lib-terminal"]
```

**Reply by gyscos** [reply]
Ah yes, I included an example in simple triple-ticks, which rustc treats as rust code by default (and obviously it doesn't compile).

Unfortunate about blt not compiling. Does it require some extra lib to build you don't have installed? I guess it might require some X11 libs that would not be on headless systems?

**Reply by correabuscar** [reply]
> Unfortunate about blt not compiling. Does it require some extra lib to build you don't have installed? I guess it might require some X11 libs that would not be on headless systems?

yes, that's it, I'm missing that lib, and apparently isn't available on Gentoo, but may be in unofficial overlays.
```rust
  = note: /usr/lib/gcc/x86_64-pc-linux-gnu/14/../../../../x86_64-pc-linux-gnu/bin/ld: cannot find -lBearLibTerminal: No such file or directory
          collect2: error: ld returned 1 exit status


error: could not compile `cursive` (example "theme_editor") due to 1 previous error
```

**Reply by gyscos** [reply]
Ah right. Yeah that's probably why I used a whitelist of features for CI as well.
